goth_tests/domain/ya-provider/test_provider_multi_activity.py

```diff
@@ -91,3 +91,189 @@ async def test_provider_multi_activity(
 
         # Payment
         await pay_all(requestor, agreement_providers)
+
+
+@pytest.mark.asyncio
+async def test_provider_renegotiate_proposal(
+    common_assets: Path,
+    config_overrides: List[Override],
+    log_dir: Path,
+):
+    """Tests providers' ability of renegotiating previously rejected proposal."""
+
+    nodes = [
+        {"name": "requestor-1", "type": "Requestor"},
+        {"name": "requestor-2", "type": "Requestor"},
+        {"name": "provider-1", "type": "VM-Wasm-Provider", "use-proxy": True},
+    ]
+    config_overrides.append(("nodes", nodes))
+
+    goth_config = load_yaml(common_assets / "goth-config.yml", config_overrides)
+
+    runner = Runner(
+        base_log_dir=log_dir,
+        compose_config=goth_config.compose_config,
+        web_root_path=common_assets / "web-root",
+    )
+
+    async with runner(goth_config.containers):
+        requestor1, requestor2 = runner.get_probes(probe_type=RequestorProbe)
+        providers = runner.get_probes(probe_type=ProviderProbe)
+
+        def build_demand(requestor):
+            return (
+                DemandBuilder(requestor)
+                .props_from_template(wasi_task_package)
+                .property("golem.srv.caps.multi-activity", True)
+                .constraints(
+                    "(&(golem.com.pricing.model=linear)\
+                    (golem.srv.caps.multi-activity=true)\
+                    (golem.runtime.name=wasmtime))"
+                )
+                .build()
+            )
+
+        async def negotiate_begin(requestor, demand, providers):
```

Comment:
My initial thought here: at least some of these internal helper functions can be moved to `goth_tests/helpers/negotiation.py`.

Reply:
I thought about it, but I think that this helpers are specific to renegotiation test. The commonly used helper is `goth_tests.helpers.negotiation.negotiate_agreements()`. I can't think of a common use for those custom helpers.